lib/cn/kblock_builder.c

```diff
@@ -1214,39 +1096,14 @@ kbb_finish(struct kblock_builder *bld, struct blk_list *kblks, u64 seqno_min, u6
      * this flag prevents other operations.
      */
     bld->finished = true;
-    bld->seqno_min = seqno_min;
-    bld->seqno_max = seqno_max;
-
-    /* Must have a spot to keep the hlog */
-    assert(bld->finished_kblks.n_blks == 0 || !kblock_is_empty(&bld->curr));
-
-    /* Finish main wbtree. If there's enough space left in the kblock, add
-     * ptree to this kblock. If not, add the ptree to the next kblock.
-     */
-    if (!kblock_is_empty(&bld->curr)) {
-        struct wbb *pt = 0;
-        uint64_t    kbsize = bld->max_size;
-        uint64_t    ptsize = (wbb_page_cnt_get(bld->ptree)) * PAGE_SIZE;
-        uint64_t    kbused =
-            (KBLOCK_HDR_PAGES + HLOG_PGC + bld->curr.blm_pgc + wbb_page_cnt_get(bld->curr.wbtree)) *
-            PAGE_SIZE;
-
-        /* Write ptree here if we have enough space */
-        if ((kbused + ptsize < kbsize)) {
-            pt_kblock = false;
-            pt = bld->ptree;
-        }
 
-        err = kblock_finish(bld, pt, bld->kvset_hlog);
-        if (ev(err))
-            return err;
-    }
+    /* In the event we have no keys, return no kblocks to the caller. */
+    if (bld->curr.num_keys == 0)
```

Comment:
Is it possible for bld->finished_kblks.n_blks to be greater than 0 and bld->curr.num_keys == 0, i.e., there are no keys in the current kblock but there are finished kblocks in this kbb? It doesn't seem to be a possibility to me.

From the code I see that the last kblock in a kbb is always written out from kbb_finish() whereas the previous kblocks are written out in kbb_add_entry(). If that's true, then bld->curr.num_keys == 0 does imply that there are no keys in this kbb and what you have here is correct. Please assert for bld->finished_kblks.n_blks == 0 in this if block.

@alexttx, please confirm...

Reply:
I believe that is correct.  An assert is good idea.